Drift kinetic pusher v.2.0.

OlegMorozovNSU opened the pull request:
1) Interpolation for drift-kinetic has been implemented (drift_kinetic_implicit.cpp). Inheritance from implicit_esirkepov with the addition of simple_interpolation for B and gradB.
2) In implicit_esirkepov, the “setup” method is made “virtual”.
3) Modified drift_kinetic_push and changed "set_fields_callback" call.
4) Added DK tests on the grid (drift_kinetic_grid): ex1 - interpolation check, ex2-4 - correspond to drift_kinetic_push.
5) Added comparison of boris_push by analytical fields and drift_kinetic_push on the grid (drift_kinetic_grid_boris): ex1 corresponds to test ex4, and ex2 corresponds to tests ex8-9.     
6) All tests ex6-7 have been removed, except for ex7_3.  Test ex7_3 has been renamed to ex5.

OlegMorozovNSU committed:
add drift_kinetic_implicit and tests on grid

OlegMorozovNSU committed:
Resolve modify/delete: keep master versions of drift_kinetic_push tests

OlegMorozovNSU committed:
merge master

vakurshakov commented on .gitignore:
```diff
@@ -7,6 +7,7 @@ external/petsc
 build/
 doc/build/
 
+tests/drift_kinetic_push/old/**
```
This shouldn't be placed into `.gitignore`, if it isn't a designed output; see for example "tests/\*\*/output/\*\*" below. In this case, you should simply remove the old tests or combine them with the existing ones, if needed.

vakurshakov commented on src/algorithms/implicit_esirkepov.h:
```diff
@@ -52,7 +53,7 @@ class ImplicitEsirkepov {
     PetscReal cache[shm];
 
-    void setup(const Vector3R& rn, const Vector3R& r0, Type t);
+    virtual void setup(const Vector3R& rn, const Vector3R& r0, Type t);
```
This `virtual` keyword is unneeded, `setup()` is not overridden

vakurshakov commented on tests/drift_kinetic_push/CMakeLists.txt:
```diff
@@ -3,14 +3,18 @@ cmake_minimum_required(VERSION 3.10)
+  drift_kinetic_push_ex5
   drift_kinetic_push_ex8
+  drift_kinetic_push_ex9
+  drift_kinetic_push_ex9_comparison
```
`build.sh` fails the compilation step, merge conflict was underresolved. Both `drift_kinetic_push_ex9` and `drift_kinetic_push_ex9_comparison` are unneeded here, this test is compiled and added below, at line 33

vakurshakov commented on src/algorithms/implicit_esirkepov.h:
```diff
@@ -14,7 +14,8 @@ class ImplicitEsirkepov {
     const Vector3R& rn, const Vector3R& r0);
 
-private:
+//private:
```
Remove the comment

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+#include "src/interfaces/particles.h"
+
+#include <algorithm>
+#include <cmath>
```
There is no need to include them, they are included through a precompiled header "src/pch.h"

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_push_ex1.cpp:
```suggestion
push.set_qm(q / m);
push.set_mp(m);
```

change `constexpr PetscReal q = -1.0` in "./tests/drift_kinetic_push/drift_kinetic_push.h", if needed

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_push_ex2.cpp:
```suggestion
push.set_qm(q / m);
push.set_mp(m);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_push_ex3.cpp:
```diff
@@ -0,0 +1,68 @@
+
+  DriftKineticPush push;
+  push.set_qm(1.0);
+  push.set_mp(1.0);
```
```suggestion
push.set_qm(q / m);
push.set_mp(m);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_push_ex4.cpp:
```diff
@@ -1,17 +1,28 @@
+constexpr PetscReal B_max = 4.0;
+constexpr PetscReal L = 10.0;
+constexpr PetscReal q = 1.0;
+constexpr PetscReal m = 1.0;
```
`build.sh` fails the compilation step here, merge conflict was underresolved. This constants conflict with those defined in "src/tests/drift_kinetic_push/drift_kinetic_push.h"

Repetitive mirror setup logic is placed mostly into the `quadratic_magnetic_mirror` namespace in "src/tests/drift_kinetic_push/drift_kinetic_push.h:20", reuse this logic here.

vakurshakov commented on src/algorithms/drift_kinetic_implicit.h:
```diff
@@ -0,0 +1,22 @@
+public:
+  DriftKineticEsirkepov(Vector3R*** E_g, Vector3R*** B_g, Vector3R*** J_g, Vector3R*** gradB_g);
+
+  // Расширенная интерполяция с градиентом B
```
1) Use English within codebase
2) Comments in this file are excessive, the code is self-explanatory

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_push_ex4.cpp:
```diff
@@ -1,17 +1,28 @@
+void get_fields(
+  const Vector3R&, const Vector3R& rn, Vector3R&, Vector3R& B_p, Vector3R& gradB_p)
 {
-  B_p = B0 + r.elementwise_product(gradB0);
-  gradB_p = gradB0;
+  auto get_B = [](PetscReal z) {
+    return B_min + (B_max - B_min) * (z * z) / (L * L);
+  };
+
+  auto get_gradB = [](PetscReal z) {
+    return 2.0 * (B_max - B_min) * z / (L * L);
+  };
+
+  B_p = Vector3R{0, 0, get_B(rn.z())};
+  gradB_p = Vector3R{0, 0, get_gradB(rn.z())};
 }
```
```suggestion
using namespace quadratic_magnetic_mirror;

void get_fields(
  const Vector3R&, const Vector3R& rn, Vector3R&, Vector3R& B_p, Vector3R& gradB_p)
{
  B_p = Vector3R{0, 0, get_B(rn.z())};
  gradB_p = Vector3R{0, 0, get_gradB(rn.z())};
}
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_push_ex4.cpp:
```diff
@@ -22,46 +33,38 @@ int main(int argc, char** argv)
   constexpr Vector3R r0(0.0, 0.0, 0.0);
-  constexpr Vector3R v0(1.0, 0.0, 1.0);
+  constexpr PetscReal v_perp = 1.0;
+  constexpr PetscReal v_par = 1.73;
```
This cannot be that way, $v \leq c$

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+  PetscCall(DMDAVecGetArrayRead(world.da, gradB_vec, &gradB_arr));
+
+  std::unique_ptr<DriftKineticEsirkepov> esirkepov =
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
```
Use of pointers here is an overkill, simply

```suggestion
  DriftKineticEsirkepov esirkepov(E_arr, B_arr, nullptr, gradB_arr);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Expected:     (%.6f, %.6f, %.6f)\n",
+                       REP3_A(gradB_expected)));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Interpolated: (%.6f, %.6f, %.6f)\n",
+                       REP3_A(gradB_p)));
```
1) This prints are excessive, can be removed or placed into `#if 0`
2) Use `LOG()` macro (same as `std::cout << std::format() << "\n"`) for logging instead of explicit calls to `PetscPrintf()`
3) Put logging statements in one line, even those exceeding line width barrier

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+    "Gradient B field interpolation failed. Expected: (%.8e %.8e %.8e), got: (%.8e %.8e %.8e)",
+    REP3_A(gradB_expected), REP3_A(gradB_interpolated));
+
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "\nAll interpolation tests passed!\n"));
```
unneeded

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+  B_p = Vector3R(0.0, 0.0, 1.0);
+
+  gradB_p = r;
+}
```
`B_p` is inconsistent with `gradB_p`, see

$$ \nabla  B_p = D_z {\bf e}_z, $$
$$ {\bf B}_p = \int  \nabla  B_p ~ d {\bf r} = \int {\bf e}_z d {\bf r} = {\bf B}_0 + r_z D_z. $$

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
+
+  Vector3R test_position_old(1, 1, 1);
+  Vector3R test_position_new(2, 2, 2);
```
positions should be within one cell

```suggestion
  Vector3R test_position_old(1, 1, 1);
  Vector3R test_position_new = test_position_old + Vector3R(0.25 * dx, 0.50 * dy, 0.75 * dz);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+
+        E_arr[k][j][i] = E_analytical;
+        B_arr[k][j][i] = B_analytical;
+        gradB_arr[k][j][i] = gradB_analytical;
```
```suggestion
        get_analytical_fields(r, E_arr[k][j][i], B_arr[k][j][i], gradB_arr[k][j][i]);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_interpolation_ex1.cpp:
```diff
@@ -0,0 +1,142 @@
+  for (PetscInt k = start[Z]; k < start[Z] + size[Z]; k++) {
+    for (PetscInt j = start[Y]; j < start[Y] + size[Y]; j++) {
+      for (PetscInt i = start[X]; i < start[X] + size[X]; i++) {
+        Vector3R r(i, j, k);
```
To map cell index to cell coordinate multiply by grid size:
```suggestion
        Vector3R r(i * dx, j * dy, k * dz);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+  PetscCall(DMDAVecRestoreArrayWrite(da, E_vec, &E_arr));
+  PetscCall(DMDAVecRestoreArrayWrite(da, B_vec, &B_arr));
+  PetscCall(DMDAVecRestoreArrayWrite(da, gradB_vec, &gradB_arr));
+
```
If fields are constant ($\nabla B_p = 0$), you can simply set each component

```suggestion
  PetscCall(VecStrideSet(E_vec, X, E0[X]));
  PetscCall(VecStrideSet(E_vec, Y, E0[Y]));
  PetscCall(VecStrideSet(E_vec, Z, E0[Z]));
  
  PetscCall(VecStrideSet(B_vec, X, B0[X]));
  PetscCall(VecStrideSet(B_vec, Y, B0[Y]));
  PetscCall(VecStrideSet(B_vec, Z, B0[Z]));
```

see https://petsc.org/release/manualpages/Vec/VecStrideSet/

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "\nConstraint violations:\n"));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Boundary violations:    %d\n", stats.boundary_violations));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Cell size violations:   %d\n", stats.cell_size_violations));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Max displacement/step:  %.8e\n", stats.max_displacement_per_step));
```
Rewrite with `LOG()`

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+  World::set_geometry(
+    geom_nx, geom_ny, geom_nz, geom_nt,
+    dx, dy, dz, dt,
+    diagnose_period);
```
`World::set_geometry()` updates global constants:

choose either

```suggestion
  World::set_geometry(20, 5, 100, 100, 1.0, 1.0, 1.0, omega_dt / B0.length(), 100);
```

or remove the duplicating logic of `World::set_geometry()`

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+
+  DriftKineticPush push_analytical;
+  push_analytical.set_qm(-1.0);
+  push_analytical.set_mp(1.0);
```
```suggestion
  push_analytical.set_qm(q / m);
  push_analytical.set_mp(m);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+  PetscCall(DMDAVecGetArrayRead(world.da, gradB_vec, &gradB_arr));
+
+  std::unique_ptr<DriftKineticEsirkepov> esirkepov =
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
```
Use of pointers here is an overkill, simply

```suggestion
  DriftKineticEsirkepov esirkepov(E_arr, B_arr, nullptr, gradB_arr);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+  constexpr Vector3R v0(0.0, 1.0, 0.0);
+  Point point_init(r0, v0);
+  PointByField point_analytical(point_init, B0, 1.0);
+  PointByField point_grid(point_init, B0, 1.0);
```
Coordinate of a gyrocenter is not the same as a coordinate of the `Point`, see `PointByField` constructor. In ex9 I've corrected the placement, this should be placed into the constructor

$$ {\bf R} = {\bf r} + [{\bf v} x {\bf B}] / \Omega,$$

where ${\bf R}$ - gyrocenter position, ${\bf r}$ - particle position, ${\bf v}$ - particle velocity, ${\bf B}$ - magnetic field, $\Omega$ - particle cyclotron frequency.

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+
+  DriftKineticPush push_grid;
+  push_grid.set_qm(-1.0);
+  push_grid.set_mp(1.0);
```
```suggestion
  push_analytical.set_qm(q / m);
  push_analytical.set_mp(m);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "Starting simulation with %d steps, dt = %.6e\n", geom_nt, dt));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "Domain: [%.1f, %.1f] x [%.1f, %.1f] x [%.1f, %.1f]\n",
+           domain_min.x(), domain_max.x(), domain_min.y(), domain_max.y(), domain_min.z(), domain_max.z()));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "Min cell size: %.6e\n", min_cell_size));
```
debug information, can be removed?

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp:
```diff
@@ -0,0 +1,342 @@
+    if (t % (geom_nt / 10) == 0) {
+      PetscCall(PetscPrintf(PETSC_COMM_WORLD, "Step %d/%d, max position error: %.8e\n",
+               t, geom_nt, stats.max_position_error));
+    }
```
debug information, can be removed?

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp:
```diff
@@ -0,0 +1,330 @@
+  Vector3R final_position_grid;
+  PetscReal simulation_time = 0.0;
+  PetscInt total_steps = 0;
+};
```
Move repetitive  `TestStatistics`, `initialize_grid_fields()`, `check_boundaries()`, `check_displacement()`, `compare_fields_and_update_stats()`, `print_statistics()` into namespace in "src/tests/drift_kinetic_push/drift_kinetic_push.h"

To simplify the function signatures and statistics usage, you can create an instance of `TestStatistics` globally, within namespace in "src/tests/drift_kinetic_push/drift_kinetic_push.h"

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp:
```diff
@@ -0,0 +1,330 @@
+void get_analytical_fields(const Vector3R&, const Vector3R& rn, Vector3R& E_p, Vector3R& B_p, Vector3R& gradB_p)
+{
+  E_p = E0;
+  B_p = B0 + (rn-Vector3R(2.0,2.0,2.0)).elementwise_product(gradB0);
```
why `Vector3R(2, 2, 2)` is removed here and `initialize_grid_fields()` doesn't have it?

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp:
```diff
@@ -0,0 +1,330 @@
+  World::set_geometry(
+    geom_nx, geom_ny, geom_nz, geom_nt,
+    dx, dy, dz, dt,
+    diagnose_period);
```
World::set_geometry(20, 20, 20, 100, 1.0, 1.0, 1.0, omega_dt / B0.length(), 100);

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp:
```diff
@@ -0,0 +1,330 @@
+  PetscCall(DMDAVecGetArrayRead(world.da, gradB_vec, &gradB_arr));
+
+  std::unique_ptr<DriftKineticEsirkepov> esirkepov =
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
```
Use of pointers here is an overkill, simply

```suggestion
  DriftKineticEsirkepov esirkepov(E_arr, B_arr, nullptr, gradB_arr);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp:
```diff
@@ -0,0 +1,330 @@
+
+  Vector3R domain_min(0.0, 0.0, 0.0);
+  Vector3R domain_max(geom_nx*dx, geom_ny*dy, geom_nz*dz);
+  PetscReal min_cell_size = std::min({2*dx, 2*dy, 2*dz});
```
why is it multiplied by two here?

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp:
```diff
@@ -0,0 +1,330 @@
+  PetscCall(print_statistics(stats));
+
+  PetscReal m = 1.0;
+  PetscReal q = 1.0;
```
this hides the definitions from "src/drift_kinetic_push/drift_kinetic_push.h", can be removed

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex4.cpp:
```diff
@@ -0,0 +1,228 @@
+#include "drift_kinetic_push.h"
+
+#include <algorithm>
```
can be removed

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex4.cpp:
```diff
@@ -0,0 +1,228 @@
+{
+  B_p = Vector3R{0, 0, get_B(rn.z())};
+  gradB_p = Vector3R{0, 0, get_gradB(rn.z())};
+}
```
```suggestion
using namespace quadratic_magnetic_mirror;

void get_fields(
  const Vector3R&, const Vector3R& rn, Vector3R&, Vector3R& B_p, Vector3R& gradB_p)
{
  B_p = Vector3R{0, 0, get_B(rn.z())};
  gradB_p = Vector3R{0, 0, get_gradB(rn.z())};
}
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex4.cpp:
```diff
@@ -0,0 +1,228 @@
+
+  Vector3R final_position_analytical;
+  Vector3R final_position_grid;
+};
```
move into h-file and reuse it here too

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex4.cpp:
```diff
@@ -0,0 +1,228 @@
+
+        gradB_arr[k][j][i].x() = 0.0;
+        gradB_arr[k][j][i].y() = 0.0;
+        gradB_arr[k][j][i].z() = get_gradB(z_bz);
```
```suggestion
        Vector3R r(i * dx, j * dy, k * dz);
        get_analytical_fields(r, E_arr[k][j][i], B_arr[k][j][i], gradB_arr[k][j][i]);
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex4.cpp:
```diff
@@ -0,0 +1,228 @@
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+PetscErrorCode initialize_grid_fields(DM da, const Vector3R& dr, Vec E_vec, Vec B_vec, Vec gradB_vec)
```
`dx`, `dy`, `dz` are global constants, there is no need to pass `dr` explicitly; if vector-like grid spacing is needed, there is `Dr[3]` too

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_ex4.cpp:
```diff
@@ -0,0 +1,228 @@
+  PetscCall(DMDAVecGetArrayRead(world.da, gradB_vec, &gradB_arr));
+
+  std::unique_ptr<DriftKineticEsirkepov> esirkepov =
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
```
no pointer here

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+  E_p = {0.0, 0.0, 0.0};
+  B_p = get_B_vector(rn);
+  gradB_p = get_gradB_vector(rn);
+}
```
Merge the new parts with the logic in `quadratic_magnetic_mirror` namespace in "src/tests/drift_kinetic_push/drift_kinetic_push.h"

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+
+  Vector3R final_position_analytical;
+  Vector3R final_position_grid;
+};
```
reuse the stats placed into "src/tests/drift_kinetic_push/drift_kinetic_push.h"

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+}
+
+PetscErrorCode initialize_grid_fields(
+  DM da, const Vector3R& dr, Vec E_vec, Vec B_vec, Vec gradB_vec)
```
no need to pass `dr`

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+        const Vector3R gradBz_vec = get_gradB_vector(pos_bz);
+        gradB_arr[k][j][i].x() = gradBx_vec.x();
+        gradB_arr[k][j][i].y() = gradBy_vec.y();
+        gradB_arr[k][j][i].z() = gradBz_vec.z();
```
use analytical field setter

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+  PetscCall(DMDAVecGetArrayRead(world.da, gradB_vec, &gradB_arr));
+
+  std::unique_ptr<DriftKineticEsirkepov> esirkepov =
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
```
no pointer here

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+    PetscCall(trace_grid.diagnose(t));
+    PetscCall(trace_boris.diagnose(t));
+
+    std::cout << "Аналитические поля" << std::endl;
```
1) Use English within codebase
2) If it is a debug information, remove it; if it is needed in some form rewrite it with `LOG()` statements

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
```diff
@@ -0,0 +1,421 @@
+    const PetscReal z_error = std::abs(point_grid.r.z() - point_boris.r.z());
+    boris_stats.max_z_error = std::max(boris_stats.max_z_error, z_error);
+    //PetscCheck(z_error < z_tolerance, PETSC_COMM_WORLD, PETSC_ERR_USER,
+    //  "Guiding center and Boris particle diverged in z: %.6e", z_error);
```
Do not leave commented code, either make it work or remove the comments

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+  E_p = {0.0, 0.0, 0.0};
+  B_p = get_B_vector(rn);
+  gradB_p = get_gradB_vector(rn);
+}
```
```suggestion
using namespace gaussian_magnetic_mirror;

void get_analytical_fields(
  const Vector3R&, const Vector3R& rn, Vector3R& E_p, Vector3R& B_p, Vector3R& gradB_p)
{
  E_p = {0.0, 0.0, 0.0};
  B_p = get_B_vector(rn);
  gradB_p = get_gradB_vector(rn);
}
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+  PetscReal max_energy_error = 0.0;
+
+  Vector3R final_position_boris;
+};
```
Move the duplicating code into h-file and reuse it here

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+        const Vector3R gradBz_vec = get_gradB_vector(pos_bz);
+        gradB_arr[k][j][i].x() = gradBx_vec.x();
+        gradB_arr[k][j][i].y() = gradBy_vec.y();
+        gradB_arr[k][j][i].z() = gradBz_vec.z();
```
use analytical field setter here

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+  const PetscReal p_perp = point.p_perp_ref();
+  const PetscReal p_parallel = point.p_par();
+  return 0.5 * (p_perp * p_perp + p_parallel * p_parallel);
+}
```
this can also be moved into h-file

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+
+  auto esirkepov = std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
+
+  constexpr PetscReal v_perp = 0.1/std::sqrt(2.);
```
```suggestion
  constexpr PetscReal v_perp = 0.1 / M_SQRT_2;
```

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+  PointTrace trace_boris(__FILE__, id + "_boris", point_boris, geom_nt / 1000);
+
+  //const PetscReal z_limit = 2*L + 1e-2;
+  //const PetscReal r_limit = 2*L + 1e-2;
```
remove commented code

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex2.cpp:
```diff
@@ -0,0 +1,436 @@
+    //PetscCheck((point_boris.r - Vector3R{axis_x, axis_y, axis_z}).length() <= r_limit,
+    //  PETSC_COMM_WORLD, PETSC_ERR_USER,
+    //  "Particle escaped radial well. r = %.6e, allowed = %.6e",
+    //  (point_boris.r - Vector3R{axis_x, axis_y, axis_z}).length(), r_limit);
```
remove commented code

vakurshakov commented on tests/drift_kinetic_push/drift_kinetic_grid_boris_ex1.cpp:
ex1 and ex2 boris tests can be merged with ex2 and ex3 grid tests

vakurshakov reviewed (commented):
Generally the code is OK. Improvements are needed:
1) Stabilize the build process after merge conflict
2) Reuse the logic from already written namespaces in h-file
3) Put the common logic from ex1, ex2 grid_boris tests and ex2, ex3, ex4 grid tests into h-file and reuse it where possible
4) Unify the logging within tests and remove the unneeded comments
5) Position correction in `PointByField` from `Point` constructor is needed, details in ex2 grid test

vakurshakov reviewed (commented):
Generally the code is OK. Improvements are needed:
1) Stabilize the build process after merge conflict
2) Reuse the logic from already written namespaces in h-file
3) Put the common logic from ex1, ex2 grid_boris tests and ex2, ex3, ex4 grid tests into h-file and reuse it where possible
4) Unify the logging within tests and remove the unneeded comments
5) Position correction in `PointByField` from `Point` constructor is needed, details in ex2 grid test

OlegMorozovNSU committed:
Stabilize the build process after merge conflict

OlegMorozovNSU committed:
General logic added to h-file

OlegMorozovNSU committed:
Unified some logging

OlegMorozovNSU committed:
Add correction radius vector in PointByField

OlegMorozovNSU committed:
Rename tests

OlegMorozovNSU committed:
preparing for the request

OlegMorozovNSU committed:
optimization

OlegMorozovNSU commented:
### Changes:
1. The build issue has been resolved
2. The general logic (creating a grid, setting fields, logging, and releasing resources) has been moved to the h-file
3. Reconfigured the tests (combined all the grid tests with the grid_boris tests, allocated a separate branch of interpolation tests, and left only the necessary push tests)
4. Unified logging in tests, removed unnecessary comments, and moved everything to an h-file
5. Larmor center correction has been added to PointByField
6. Minor bugs with speed determination and physical field consistency have been fixed
### To-do:
1. The h-file requires logic adjustments

vakurshakov committed:
run-clang-format.sh

vakurshakov committed:
unneeded correction is removed

vakurshakov committed:
unneeded correction is removed

vakurshakov committed:
1) magnetic mirror fields are translated onto relative coordinates
2) field manipulations through context is simplified:
  a) one context for fields
  b) initialization is field set and get array read
  c) finalization is restore array read and destruction of vectors
3) stats are simplified to one structure, one print_statistics() call
4) traces are simplified to one diagnostic made of three table diagnostics

vakurshakov committed:
correction of omega_dt
